src/Microsoft.DotNet.VersionTools/BuildManifest/Model/BlobArtifactModel.cs

```diff
@@ -0,0 +1,53 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Xml.Linq;
+
+namespace Microsoft.DotNet.VersionTools.BuildManifest.Model
+{
+    public class BlobArtifactModel
+    {
+        public Dictionary<string, string> Attributes { get; set; } = new Dictionary<string, string>();
+
+        public string[] VirtualDirectorySegments { get; set; } = new string[0];
+
+        public string Id
+        {
+            get { return Attributes[nameof(Id)]; }
+            set { Attributes[nameof(Id)] = value; }
+        }
+
+        public string VirtualDirectory
+        {
+            get { return string.Join("/", VirtualDirectorySegments); }
+            set { VirtualDirectorySegments = value.Split('/'); }
+        }
+
+        public override string ToString() => $"Blob {VirtualDirectory}/{Id}";
+
+        public XElement ToXml() => new XElement(
+            "Blob",
+            Attributes.CreateXmlAttributes());
+
+        public static BlobArtifactModel Parse(XElement xml) => new BlobArtifactModel
+        {
+            VirtualDirectorySegments = ElementVirtualDirSegments(xml),
+            Attributes = xml.CreateAttributeDictionary()
+        };
+
+        private static string[] ElementVirtualDirSegments(XElement xml)
+        {
+            // Get all ancestor dirs from top to bottom.
+            IEnumerable<XElement> dirs = xml.Ancestors("Dir").Reverse();
+
+            // Get dir name strings, in order. Allow null and empty to avoid crashing if we have
+            // an element like <Dir> or <Dir Name="">. This can happen if someone passes an extra
+            // slash like "foo/bar//baz". The "//" is a valid virtual dir, even though it breaks the
+            // virtual dir metaphor.
+            return dirs.Select(dir => dir.Attribute("Name")?.Value).ToArray();
```

Comment:
What purpose do the empty dirs serve and how could we end up with them, other than bad path processing?  We could just make a hard requirement not to have empty dirs.  There doesn't seem to be a reason that they should be generated by the repos.

Reply:
Compatibility here could mean the difference between a build being weird but usable vs. tanked. With the change to the flattened approach I'm keeping the compatibility because it would take work to not be compatible, but probably worth adding a warning in `PushToBlobFeed` because it's certainly not good or needed to have empty-name virtual dirs.

Reply:
Sounds good.  I always get worried with string parsing of paths because it's so easy to make mistakes with trailing paths, reconstructing paths, etc.